src/main/java/net/apnic/rdap/error/controller/ErrorController.java

```diff
@@ -0,0 +1,190 @@
+package net.apnic.rdap.error.controller;
+
+import javax.servlet.RequestDispatcher;
+import javax.servlet.http.HttpServletRequest;
+
+import net.apnic.rdap.error.MalformedRequestException;
+import net.apnic.rdap.rdap.RDAPError;
+import net.apnic.rdap.rdap.RDAPObjectFactory;
+import net.apnic.rdap.resource.ResourceNotFoundException;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.web.ErrorAttributes;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.servlet.NoHandlerFoundException;
+
+/**
+ * Error controller for Zuul and spring that responds to requests on the error
+ * path.
+ *
+ * This is used in conjunction with the error controller. Unfotunetly to the way
+ * errors are handled in Zuul any exception thrown is redirected to the error
+ * path.
+ *
+ * Its the job of this class to take the exception and work out the most
+ * appropriate RDAP response.
+ */
+@ControllerAdvice
+@RestController
+public class ErrorController
+    implements org.springframework.boot.autoconfigure.web.ErrorController
+{
+    private static final String EXCEPTION_ATTRIBUTE =
+        "javax.servlet.error.exception";
+    private static final String ORIGIN_ATTRIBUTE = "origin";
+    private static final MediaType RDAP_MEDIA_TYPE =
+        new MediaType("application", "rdap+json");
+
+    @Value("${error.path:/error}")
+    private String errorPath;
+
+    private RDAPObjectFactory rdapObjectFactory = null;
+    private HttpHeaders responseHeaders = null;
+
+    /**
+     * Main constructor which takes the services RDAPObjectFactory to construct
+     * RDAPError's from.
+     *
+     * @param rdapObjectFactory Object factory to construct RDAPError objects
+     *                          from
+     */
+    @Autowired
+    public ErrorController(RDAPObjectFactory rdapObjectFactory)
+    {
+        this.rdapObjectFactory = rdapObjectFactory;
+        setupResponseHeaders();
+    }
+
+    /**
+     * Internal util function for checking if a throwable was caused by the
+     * provided class.
+     *
+     * Rudimentary method for unraveling the wrapping done by Zuul.
+     *
+     * @param th Throwable to check
+     * @param clazz Class to check is contained within the the throwable.
+     */
+    private boolean causedBy(Throwable th, Class<?> clazz)
```

Comment:
Very true and I was going to do it this way but the following reasons stopped me.

Spring is already there and setup for returning REST response. Doing this in every place that an error could happen would be cumbersome and require a lot manual lifting and writing code for things Spring is doing out of the box already.

On top of that we don't loose this error code it still needs to be there for any exception that happens within Zuul or a runtime exception that has not been caught by the above suggestion.

With this we just end up with 3 places for error handling logic now instead of a central management point thats easier in my opinion to maintain.

The last thought that stopped me from doing it this way originally is its going against the grain of what Zuul is doing. i.e it just a proxy redirect layer to direct traffic and not one that generates tangible responses.